ksql-engine/src/main/java/io/confluent/ksql/streams/StreamsUtil.java

```diff
@@ -27,4 +28,8 @@ static boolean useProvidedName(final KsqlConfig ksqlConfig) {
         KsqlConfig.KSQL_USE_NAMED_INTERNAL_TOPICS_ON
     );
   }
+
+  public static String buildOpName(final QueryContext opContext) {
```

Comment:
I'm not sure I understand proper usage of the query context and what defines an "op name". What information should be part of the context and what is extraneous?  Perhaps restricting the `push` interface to accept something stricter than `String` would help.

Also small nit, this can be method in `QueryContext` instead of in a static helper.

Reply:
We have to build names for things in different parts of the computed topology:
    - Some streams operations (group-by, aggregate, join) need names. (Internally those names get used to come up with a name for repartition and state-store topics. But this is kind of an implementation detail of those KStream/KTable methods - from the caller's point of view it's just a name for the operation that should be kept stable). This name must be unique within the topology, and stable across rebuilds. A good way of coming up with a stable, unique name is using a hierarchical name that's built from the logical plan node and operation inside that logical node. (as opposed to something like a counter which is much harder to keep stable. imagine if you are processing a join and tweak the query building code to process the left side before the right - all your names will change) 
    - The record processing log needs names for its loggers. Each logger should have a stable name within a hierarchy that can be correlated with the query execution plan.

This query context stuff gives us a pattern we can use to build these hierarchical names. The context is where we currently are in the hierarchy, and we can add levels as we go into specific parts of the topology.

> Also small nit, this can be method in QueryContext instead of in a static helper.

I had it this way at first, but it started to feel like the specific ways QueryContext was being used were leaking into QueryContext itself.

Reply:
Thanks for the details, definitely helps me get a deeper understanding of this PR. I'm still a little hesitant to communicate sensitive (i.e. must be stable) information through unstructured strings, but I don't have any better suggestions...

> I had it this way at first, but it started to feel like the specific ways QueryContext was being used were leaking into QueryContext itself.

Fair enough!

Reply:
> I'm still a little hesitant to communicate sensitive (i.e. must be stable) information through unstructured strings

We do have very rigorous testing that will verify that this doesn't change release after release. Basically, we generate topologies from all the query-validation tests (QueryTranslationTest.java) and write them out on every release. Then on every build we generate new topologies and make sure they match the persisted ones.